version: '3.8'

services:
  django:
    container_name: 0xfaustio_django
    build: .
    volumes:
      - .:/opt/services/django
      - static_volume:/opt/services/django/src/static
      - media_volume:/opt/services/django/src/media
      - keras_volume:/opt/services/django/src/models
    env_file:
      - .env
    environment:
      - DD_AGENT_HOST=0xfaustio_datadog
      - DD_TRACE_AGENT_PORT=8126
      - DD_ANALYTICS_ENABLED=true
      - DD_TRACE_ANALYTICS_ENABLED=true
    networks:
      - datadog_network
    depends_on:
      - postgres
    labels:
      com.datadoghq.ad.logs: '[{"source": "django", "service": "web application"}]'
  
  nginx:
    container_name: 0xfaustio_nginx
    image: nginx:1.18.0
    volumes:
      - ./config/nginx/conf.d:/etc/nginx/conf.d
      - static_volume:/opt/services/django/src/static
      - media_volume:/opt/services/django/src/media
    depends_on:
      - django
    networks:
      - datadog_network
    ports:
      - 8000:80
    labels:
      com.datadoghq.ad.check_names: '["nginx"]'
      com.datadoghq.ad.init_configs: '[{}]'
      com.datadoghq.ad.instances: '[{"nginx_status_url": "http://%%host%%:81/nginx_status"}]'
      com.datadoghq.ad.logs: '[{"source": "nginx", "service": "reverse proxy"}]'

  postgres:
    container_name: 0xfaustio_postgres
    image: postgres:12.3
    volumes:
      - postgres_volume:/var/lib/postgresql/data
    env_file:
      - config/postgres/.env
    networks:
      - datadog_network
    ports:
      - '5432:5432'
    labels:
      com.datadoghq.ad.check_names: '["postgres"]'
      com.datadoghq.ad.init_configs: '[{}]'
      com.datadoghq.ad.instances: '[{"host":"%%host%%", "port":5432,"username":"${DATADOG_USER}","password":"${DATADOG_PASSWORD}"}]'
      com.datadoghq.ad.logs: '[{"source": "postgresql", "service": "database"}]'

  datadog:
    container_name: 0xfaustio_datadog
    image: datadog/agent:7.19.2
    env_file:
      - config/datadog/.env
    environment: 
      - DD_LOGS_ENABLED=true
      - DD_LOGS_CONFIG_CONTAINER_COLLECT_ALL=true
      - DD_APM_ENABLED=true
      - DD_APM_NON_LOCAL_TRAFFIC=true
      - DD_DOGSTATSD_NON_LOCAL_TRAFFIC=true
      - DD_DOGSTATSD_SOCKET=/var/run/datadog/dsd.socket
      - DD_DOGSTATSD_ORIGIN_DETECTION=true
    networks:
      - datadog_network
    ports:
      - 8126:8126/tcp
    volumes:
      - /var/run/docker.sock:/var/run/docker.sock
      - /proc/:/host/proc/:ro
      - /sys/fs/cgroup:/host/sys/fs/cgroup:ro
      - /var/socket/run/datadog:/var/run/datadog
      - ./config/datadog/keras.d/conf.yaml:/etc/datadog-agent/conf.d/keras.d/conf.yaml
      - keras_volume:/var/lib/keras/
    labels:
      com.datadoghq.ad.logs: '[{"source": "agent", "service": "datadog agent"}]'
    pid: host
    depends_on:
      - django
      - nginx
      - postgres
      - keras

  keras:
    container_name: 0xfaustio_keras
    build: ./nn
    environment:
      - TZ=Europe/Dublin
    stdin_open: true
    tty: true
    volumes:
      - keras_volume:/opt/services/keras/models
      - /var/socket/run/datadog:/var/run/datadog:ro
    networks:
      - datadog_network
    labels:
      com.datadoghq.ad.logs: '[{"source": "keras", "service": "machine learning"}]'

networks:
  datadog_network:
    driver: bridge  

volumes:
  postgres_volume:
  static_volume:
  media_volume:
  keras_volume: